File: packages/orchestra/src/modules/agent-runner.ts
export class AgentRunner {
    private graph: any;
    private state: Record<string, any>;
    private registry: any;
    constructor(graph: any, registry: any) {
        this.graph = graph;
        this.registry = registry;
        this.state = {};
    }
    async run(inputs: Record<string, any>) {
        this.state.inputs = inputs;
        let nodeId = this.graph.start;
        let safety = 0;
        let output: any = null;
        if (nodeId) {
            while (nodeId && safety < 500) {
                safety += 1;
                const node = this.graph.nodes[nodeId];
                const res = await this.executeNode(nodeId, node);
                nodeId = this.resolveNext(node, res);
                output = res;
            }
        } else {
            output = { ok: false, error: { code: "missing_start", message: "Graph has no start node" } };
        }
        return { state: this.state, last: output };
    }
    private async executeNode(nodeId: string, node: any) {
        const ctx = { inputs: this.state.inputs, state: this.state, nodeId, graphId: this.graph.id, graph: this.graph };
        let res: any = null;
        if (node.type === "planner") {
            const planned = await this.registry.plan(ctx, {
                tools: node.tools ?? [],
                outputSchema: node.output_schema ?? null,
            });
            this.applyEmit(node.emit, planned);
            res = { ok: true, result: planned };
        } else if (node.type === "executor") {
            const resolvedInput = this.resolveTemplates(node.run?.input);
            if (node.run?.op === "api.call") {
                const called = await this.registry.callApi(ctx, {
                    target: node.run.target,
                    input: resolvedInput,
                });
                if (called?.ok === true) {
                    this.applyEmit(node.emit, called);
                }
                res = called;
            } else {
                if (node.run?.op === "state.select") {
                    const selected = this.runStateSelect(resolvedInput);
                    if (selected?.ok === true) {
                        this.applyEmit(node.emit, selected);
                    }
                    res = selected;
                } else {
                    throw new Error(`Unknown executor op: ${node.run?.op}`);
                }
            }
        } else if (node.type === "control") {
            const decided = this.evalBranch(node.condition);
            res = { ok: true, result: decided };
        } else if (node.type === "terminal") {
            if (node.emit) {
                const v = this.resolveInputTemplate(node.emit);
                this.state.output = v;
            } else {
                this.state.output = this.state.output ?? null;
            }
            res = { ok: true, result: this.state.output };
        } else {
            res = { ok: false, error: { code: "unknown_node_type", message: String(node.type) } };
        }
        return res;
    }
    runStateSelect(input: any) {
        const source = input.from;
        console.log("first", source);
        if (!Array.isArray(source)) {
            throw new Error("state.select: source is not an array");
        }
        console.log(source);
        let items = source;
        if (input.filter) {
            items = items.filter((item: any) => item?.[input.filter.field] === input.filter.equals);
        }
        const item = items[input.index ?? 0];
        if (!item) {
            throw new Error("state.select: index out of bounds");
        }
        if (!input.field || !(input.field in item)) {
            throw new Error("state.select: field not found");
        }
        return { ok: true, result: item[input.field] };
    }

    private resolveNext(node: any, res: any) {
        let next: any = null;
        if (node.type === "control") {
            next = res?.result?.next ?? null;
        } else if (typeof node.next === "string") {
            next = this.interpolateNext(node.next, res?.result);
        } else if (node.on) {
            if (res?.ok === true && node.on.ok) {
                next = node.on.ok;
            } else if (res?.ok === false && node.on.error) {
                next = node.on.error;
            } else {
                next = null;
            }
        } else {
            next = null;
        }
        return next;
    }
    private applyEmit(emit: any, payload: any) {
        if (emit && payload) {
            for (const dest of Object.keys(emit)) {
                const src = emit[dest];
                const key = dest.startsWith("state.") ? dest.slice(6) : dest;
                if (typeof src === "string") {
                    if (src.startsWith("${") && src.endsWith("}")) {
                        this.state[key] = this.getPath(src.slice(2, -1));
                    } else if (src === "." || src === "$self") {
                        this.state[key] = payload;
                    } else {
                        this.state[key] = payload[src];
                    }
                } else {
                    this.state[key] = src;
                }
            }
        }
    }
    resolveTemplates(value: any): any {
        if (typeof value === "string") {
            return this.resolveInputTemplate(value);
        }
        if (Array.isArray(value)) {
            return value.map((v) => this.resolveTemplates(v));
        }
        if (value && typeof value === "object") {
            const out: any = {};
            for (const [k, v] of Object.entries(value)) {
                out[k] = this.resolveTemplates(v);
            }
            return out;
        }
        return value;
    }
    private resolveInputTemplate(tpl: any): any {
        if (tpl === null || tpl === undefined) {
            return tpl;
        } else if (Array.isArray(tpl)) {
            return tpl.map((x) => this.resolveInputTemplate(x));
        } else if (typeof tpl === "object") {
            const out: any = {};
            for (const k of Object.keys(tpl)) {
                out[k] = this.resolveInputTemplate(tpl[k]);
            }
            return out;
        } else if (typeof tpl === "string" && tpl.startsWith("${") && tpl.endsWith("}")) {
            return this.getPath(tpl.slice(2, -1));
        } else {
            return tpl;
        }
    }
    private interpolateNext(next: string, planned: any) {
        let out = next;
        if (next.startsWith("${") && next.endsWith("}")) {
            const key = next.slice(2, -1);
            if (planned && planned[key] !== undefined) {
                out = String(planned[key]);
            } else {
                out = null as any;
            }
        }
        return out;
    }
    private evalBranch(condition: any) {
        let next: string | null = null;
        if (condition?.op === "control.branch" && Array.isArray(condition.cases)) {
            for (const c of condition.cases) {
                const path = c?.when?.path;
                const equals = c?.when?.equals;
                if (typeof path === "string") {
                    const v = this.getPath(path);
                    if (v === equals) {
                        next = String(c.next);
                        break;
                    } else {
                        next = next;
                    }
                } else {
                    next = next;
                }
            }
            if (!next) {
                next = condition.default ? String(condition.default) : null;
            } else {
                next = next;
            }
        } else {
            next = null;
        }
        return { next };
    }
    private getPath(path: string) {
        const parts = path.replace(/^state\./, "").split(".");
        let cur: any = this.state;
        for (const p of parts) {
            if (cur && typeof cur === "object" && p in cur) {
                cur = cur[p];
            } else {
                cur = undefined;
            }
        }
        return cur;
    }
}

File: packages/polaris/src/python/agent_runner.py
class AgentRunner:
    def __init__(self, graph, registry):
        self.graph = graph
        self.registry = registry
        self.state = {}

    async def run(self, inputs):
        self.state["inputs"] = inputs
        node_id = self.graph.get("start")
        safety = 0
        output = None
        if node_id:
            while node_id and safety < 500:
                safety += 1
                node = self.graph["nodes"][node_id]
                res = await self.execute_node(node_id, node)
                node_id = self.resolve_next(node, res)
                output = res
        else:
            output = {
                "ok": False,
                "error": {
                    "code": "missing_start",
                    "message": "Graph has no start node"
                }
            }
        return {"state": self.state, "last": output}

    async def execute_node(self, node_id, node):
        ctx = {
            "inputs": self.state.get("inputs"),
            "state": self.state,
            "nodeId": node_id,
            "graphId": self.graph.get("id"),
            "graph": self.graph
        }
        res = None
        if node["type"] == "planner":
            planned = await self.registry.plan(ctx, {
                "node": node,
                "prompt": node.get("prompt", ""),
                "tools": node.get("tools", []),
                "outputSchema": node.get("output_schema")
            })
            self.apply_emit(node.get("emit"), planned)
            res = {"ok": True, "result": planned}
        elif node["type"] == "executor":
            resolved_input = self.resolve_templates(node.get("run", {}).get("input"))
            op = node.get("run", {}).get("op")
            if op == "api.call":
                called = await self.registry.call_api(ctx, {
                    "target": node["run"]["target"],
                    "input": resolved_input
                })
                if called.get("ok") is True:
                    self.apply_emit(node.get("emit"), called)
                res = called
            elif op == "state.select":
                selected = self.run_state_select(resolved_input)
                if selected.get("ok") is True:
                    self.apply_emit(node.get("emit"), selected)
                res = selected
            elif op == "state.lookup":
                input_val = resolved_input
                source = input_val.get("from")
                if not isinstance(source, list):
                    raise Exception("state.lookup: source is not an array")
                match = None
                for item in source:
                    if item.get(input_val.get("match", {}).get("field")) == input_val.get("match", {}).get("equals"):
                        match = item
                        break
                if not match:
                    raise Exception("state.lookup: no match")
                select = input_val.get("select")
                if select not in match:
                    raise Exception("state.lookup: select field not found")
                res = {"ok": True, "result": match[select]}
                self.apply_emit(node.get("emit"), res)
            else:
                raise Exception(f"Unknown executor op: {op}")
        elif node["type"] == "control":
            decided = self.eval_branch(node.get("condition"))
            res = {"ok": True, "result": decided}
        elif node["type"] == "terminal":
            if node.get("emit"):
                v = self.resolve_input_template(node["emit"])
                self.state["output"] = v
            else:
                self.state["output"] = self.state.get("output")
            res = {"ok": True, "result": self.state.get("output")}
        else:
            res = {
                "ok": False,
                "error": {
                    "code": "unknown_node_type",
                    "message": str(node.get("type"))
                }
            }
        return res

    def run_state_select(self, input_val):
        source = input_val.get("from")
        if not isinstance(source, list):
            raise Exception("state.select: source is not an array")
        items = source
        filt = input_val.get("filter")
        if filt:
            items = [i for i in items if i.get(filt.get("field")) == filt.get("equals")]
        index = input_val.get("index", 0)
        if index >= len(items):
            raise Exception("state.select: index out of bounds")
        item = items[index]
        field = input_val.get("field")
        if not field or field not in item:
            raise Exception("state.select: field not found")
        return {"ok": True, "result": item[field]}

    def resolve_next(self, node, res):
        next_val = None
        if node["type"] == "control":
            next_val = res.get("result", {}).get("next")
        elif isinstance(node.get("next"), str):
            next_val = self.interpolate_next(node["next"], res.get("result"))
        elif node.get("on"):
            if res.get("ok") is True and node["on"].get("ok"):
                next_val = node["on"]["ok"]
            elif res.get("ok") is False and node["on"].get("error"):
                next_val = node["on"]["error"]
        return next_val

    def apply_emit(self, emit, payload):
        if emit and payload:
            for dest, src in emit.items():
                key = dest[6:] if dest.startswith("state.") else dest
                if isinstance(src, str):
                    if src.startswith("${") and src.endswith("}"):
                        self.state[key] = self.get_path(src[2:-1])
                    elif src in (".", "$self"):
                        self.state[key] = payload
                    else:
                        self.state[key] = payload.get(src)
                else:
                    self.state[key] = src

    def resolve_templates(self, value):
        if isinstance(value, str):
            return self.resolve_input_template(value)
        if isinstance(value, list):
            return [self.resolve_templates(v) for v in value]
        if isinstance(value, dict):
            return {k: self.resolve_templates(v) for k, v in value.items()}
        return value

    def resolve_input_template(self, tpl):
        if tpl is None:
            return tpl
        if isinstance(tpl, list):
            return [self.resolve_input_template(x) for x in tpl]
        if isinstance(tpl, dict):
            return {k: self.resolve_input_template(v) for k, v in tpl.items()}
        if isinstance(tpl, str) and tpl.startswith("${") and tpl.endswith("}"):
            return self.get_path(tpl[2:-1])
        return tpl

    def interpolate_next(self, next_val, planned):
        if next_val.startswith("${") and next_val.endswith("}"):
            key = next_val[2:-1]
            return str(planned.get(key)) if planned and key in planned else None
        return next_val

    def eval_branch(self, condition):
        next_val = None
        if condition and condition.get("op") == "control.branch":
            for c in condition.get("cases", []):
                path = c.get("when", {}).get("path")
                equals = c.get("when", {}).get("equals")
                if isinstance(path, str):
                    v = self.get_path(path)
                    if v == equals:
                        next_val = str(c.get("next"))
                        break
            if not next_val:
                default = condition.get("default")
                next_val = str(default) if default else None
        return {"next": next_val}

    def get_path(self, path):
        parts = path.replace("state.", "").split(".")
        cur = self.state
        for p in parts:
            if isinstance(cur, dict) and p in cur:
                cur = cur[p]
            else:
                return None
        return cur
